### PR TITLE
Contract high-level module

### LINKED ISSUE #21: Contract high level module
- deploy
- call
- call_static
- decode

### PR DESCRIPTION
**Pull request by cheezus1**
closes #21, #26

### CODE REVIEW

File: apps/core/lib/contract.ex [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+      iex> network_id = "ae_uat"
+      iex> url = "https://sdk-testnet.aepps.com/v2"
+      iex> internal_url = "https://sdk-testnet.aepps.com/v2"
+      iex> client = Core.Client.new(%{pubkey: pubkey, privkey: privkey}, network_id, url, internal_url)
+      iex> source_code = "contract Number =\\n  record state = { number : int }\\n\\n  function init(x : int) =\\n    { number = x }\\n\\n  function add_to_number(x : int) = state.number + x"
+      iex> init_args = "(42)"
+      iex> Core.Contract.deploy(client, source_code, init_args)
+      {:ok, "ct_2sZ43ScybbzKkd4iFMuLJw7uQib1dpUB8VDi9pLkALV5BpXXNR"}
+  """
+
```
[the remark this answers is not in the document]
Comment (cheezus1) [reply apps/core/lib/contract.ex]: > why the new line?

auto formatter did it this way

File: apps/utils/lib/encoding.ex [hunk 2]
```diff
@@ -72,6 +72,12 @@ defmodule Utils.Encoding do
     data
   end
 
+  @doc """
+  false
+  """
+  @spec prefix_encode_base64(String.t(), binary()) :: base64()
```
[the remark this answers is not in the document]
Comment (cheezus1) [reply apps/utils/lib/encoding.ex]: we thought that this function wouldn't be called directly

File: apps/utils/lib/encoding.ex [hunk 3]
```diff
@@ -96,6 +102,35 @@ defmodule Utils.Encoding do
     |> Base.encode64()
   end
 
+  @doc """
+  false
```
[the remark this answers is not in the document]
Comment (cheezus1) [reply apps/utils/lib/encoding.ex]: same as above

File: apps/utils/lib/encoding.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (thepiwo) [reply apps/utils/lib/encoding.ex]: no matter, if it is public it can be called directly and should be documented

File: apps/utils/lib/transaction.ex [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    else
+      {:ok, %Error{reason: message}} ->
+        {:error, message}
+
+      {:error, %Env{} = env} ->
+        {:error, env}
+    end
+  end
+
+  defp await_mining(connection, tx_hash) do
```
[the remark this answers is not in the document]
Comment (d-velev) [reply apps/utils/lib/transaction.ex]: We could have async and non-async version of a function, can do it at any time

File: apps/utils/lib/keys.ex [hunk 5]
```diff
@@ -62,6 +62,12 @@ defmodule Utils.Keys do
   @spec sign(binary(), binary()) :: signature()
   def sign(message, privkey), do: :enacl.sign_detached(message, privkey)
 
+  @doc """
+  false
```
[the remark this answers is not in the document]
Comment (cheezus1) [reply apps/utils/lib/keys.ex]: same as above

**Comment by d-velev** [comment]
@thepiwo The functions / modules that have doc false are one that we've decided not to have as part of the public API, do you think those that we've left private would be useful?

File: apps/utils/lib/transaction.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (thepiwo) [reply apps/utils/lib/transaction.ex]: sounds good

**Comment by thepiwo** [comment]
@d-velev I don't think there is a good reason to keep utilities private

**Comment by d-velev** [comment]
@thepiwo Well for example the Utils.Transaction.post function expects a list of fields to be serialized to RLP, would the user have to lookup the serialization template from the protocol repo or the node repo to use it?  It doesn't make sense for us to worry about documenting them well enough so that people can use them and having them public will make users think they should be used for something

**Comment by thepiwo** [comment]
If their application wants to do some of these steps on its own, they could do it using this endpoint

**Comment by DanielaIvanova** [comment]
Is this PR resolves #26?

**Comment by d-velev** [comment]
@DanielaIvanova yes, will update description

**Comment by DanielaIvanova** [comment]
@d-velev The problem with `AeternityNode.Model.SophiaJsonData` model is resolved as we are not using Sophia related endpoints. Could you adjust the description, please.

**Comment by thepiwo** [comment]
I assume this will need to be adjusted to fit the new serializations?

**Comment by DanielaIvanova** [comment]
@thepiwo The serialization can be done in both ways: with already prepared fields and the transaction.

File: apps/core/lib/contract.ex [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+      iex> network_id = "ae_uat"
+      iex> url = "https://sdk-testnet.aepps.com/v2"
+      iex> internal_url = "https://sdk-testnet.aepps.com/v2"
+      iex> client = Core.Client.new(%{pubkey: pubkey, privkey: privkey}, network_id, url, internal_url)
+      iex> contract_address = "ct_2sZ43ScybbzKkd4iFMuLJw7uQib1dpUB8VDi9pLkALV5BpXXNR"
+      iex> function_name = "add_to_number"
+      iex> function_args = "33"
+      iex> {:ok, %{return_value: data, return_type: "ok"}} = Core.Contract.call(client, contract_address, function_name, function_args)
+      iex> data_type = "int"
+      iex> Core.Contract.decode_return_value(client, data_type, data)
```
Comment (thepiwo) [inline apps/core/lib/contract.ex]: documentation seems incorrect

File: apps/core/lib/contract.ex [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  defp hash(payload) do
+    :enacl.generichash(@hash_bytes, payload)
+  end
+
+  defp compile(source_code) do
```
Comment (thepiwo) [inline apps/core/lib/contract.ex]: I assume compiling should be exposed

File: apps/core/lib/contract.ex [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+  ## Examples
+      iex> pubkey = "ak_6A2vcm1Sz6aqJezkLCssUXcyZTX7X8D5UwbuS2fRJr9KkYpRU"
+      iex> privkey = "a7a695f999b1872acb13d5b63a830a8ee060ba688a478a08c6e65dfad8a01cd70bb4ed7927f97b51e1bcb5e1340d12335b2a2b12c8bc5221d63c4bcb39d41e61"
+      iex> network_id = "ae_uat"
+      iex> url = "https://sdk-testnet.aepps.com/v2"
+      iex> internal_url = "https://sdk-testnet.aepps.com/v2"
+      iex> client = Core.Client.new(%{pubkey: pubkey, privkey: privkey}, network_id, url, internal_url)
+      iex> source_code = "contract Number =\\n  record state = { number : int }\\n\\n  function init(x : int) =\\n    { number = x }\\n\\n  function add_to_number(x : int) = state.number + x"
+      iex> init_args = "42"
+      iex> Core.Contract.deploy(client, source_code, init_args)
```
Comment (thepiwo) [inline apps/core/lib/contract.ex]: when trying the example I receive some error

```
{:error,
 [
   [
     108,
     105,
     110,
     101,
     32,
     '"\\\\n  record state = { number : int }\\\\n\\\\n  function init(x : int) =\\\\n    { number = x }\\\\n\\\\n  function add_to_number(x : int) = state.number + x"',
     44,
     32,
     99,
     111,
     108,
     117,
     109,
     110,
     32,
     [
       123,
       [
         '1',
         44,
         10,
         [
           32,
           32,
           [
             32,
             [
               [
                 [[32, '  ', 32, 32], 32, '  ', 32, 32],
                 [32, '  ', 32, 32],
                 32,
                 '  ',
                 32,
                 32
               ],
               [[32, '  ', 32, 32], 32, '  ', 32, 32],
               [32, '  ', 32, 32],
               32,
               '  ',
               32,
               32
             ],
             [
               [[32, '  ', 32, 32], 32, '  ', 32, 32],
               [32, '  ', 32, 32],
               32,
               '  ',
               32,
               32
             ],
             [[32, '  ', 32, 32], 32, '  ', 32, 32],
             [32, '  ', 32, 32],
             32,
             '  ',
             32,
             32
           ],
           32,
           [
             [
               [[32, '  ', 32, 32], 32, '  ', 32, 32],
               [32, '  ', 32, 32],
               32,
               '  ',
               32,
               32
             ],
             [[32, '  ', 32, 32], 32, '  ', 32, 32],
             [32, '  ', 32, 32],
             32,
             '  ',
             32,
             32
           ],
           [
             [[32, '  ', 32, 32], 32, '  ', 32, 32],
             [32, '  ', 32, 32],
             32,
             '  ',
             32,
             32
           ],
           [[32, '  ', 32, 32], 32, '  ', 32, 32],
           [32, '  ', 32, 32],
           32,
           '  ',
           32,
           32
         ],
         '18'
       ],
       125
     ],
     58,
     32,
     'scan error'
   ]
 ]}
```

Probably the errormessage should be encoded somehow?

File: apps/core/lib/contract.ex [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+           TransactionUtils.post(
+             connection,
+             privkey,
+             network_id,
+             contract_create_fields,
+             :contract_create_tx
+           ),
+         contract_pubkey = compute_contract_pubkey(pubkey_binary, nonce) do
+      {:ok, contract_pubkey}
+    else
```
Comment (thepiwo) [inline apps/core/lib/contract.ex]: probably errorhandling should be improved in general, but this can be done in a later PR

File: apps/core/lib/contract.ex (hunk 8 above)
Comment (d-velev) [reply apps/core/lib/contract.ex]: I don't know how this can be encoded, it's probably better to have it like this as the error could be something useful like `['Cannot unify int\n         and list(int)\nwhen checking the assignment of the field\n  number : int (at line 5, column 14)\nto the value\n  x : list(int)\n']`. In this case there were double backslashes for some reason, you shouldn't be seeing a message like this too often

**Comment by d-velev** [comment]
@thepiwo Can we use the generated hex doc instead of having everything in the readme?

File: apps/core/lib/contract.ex (hunk 8 above)
Comment (thepiwo) [reply apps/core/lib/contract.ex]: So what do we do about this?

File: apps/utils/lib/transaction.ex [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+  defp await_mining(_connection, _tx_hash, 0),
+    do:
+      {:error,
+       "Transaction wasn't mined after #{@await_attempts * @await_attempt_interval / 1000} seconds"}
+
+  defp await_mining(connection, tx_hash, attempts) do
+    :timer.sleep(@await_attempt_interval)
+
+    case TransactionApi.get_transaction_info_by_hash(connection, tx_hash) do
```
Comment (DanielaIvanova) [inline apps/utils/lib/transaction.ex]: @d-velev Could you explain us why you are using this function instead of `TransactionApi.get_transaction_by_hash` ?

Comment (d-velev) [reply apps/utils/lib/transaction.ex]: We need the call result when calling, I've added a case for every other tx which uses that endpoint (not pushed yet)

File: apps/utils/lib/transaction.ex [hunk 11]
```diff
@@ -87,32 +87,47 @@ defmodule Utils.Transaction do
...
+        await_mining(connection, tx_hash, attempts - 1, type)
 
       {:error, %Env{} = env} ->
         {:error, env}
     end
   end
 
   # TODO: to be calculated dynamically
-  def calculate_min_fee(), do: 2_000_000_000_000_000_000
+  def calculate_min_fee(), do: 2_000_000_000_000_000
```
Comment (DanielaIvanova) [inline apps/utils/lib/transaction.ex]: Could you adjust the calculation of minimum fee, please.

Comment (d-velev) [reply apps/utils/lib/transaction.ex]: I am merging master right now, can't do it before that :smile:

**Comment by d-velev** [comment]
Tests don't pass as the testnet nodes reject the transaction (fee too low)

File: README.md [hunk 12]
```diff
@@ -1,39 +1,45 @@
...
+Functions that make requests to a node require a client as their first parameter. A client is defined like so:
+```elixir
+Core.Client.new(
+  %{
+    public: "ak_aNC4vAk5RDQdDpWdReL9ou37fBvGVYoGDah6cL1GTva3WV3JU",
+    secret:
+      "e647a7711a2ec9476033984e2805c0631730f402eae62bf23675084fd02034bc4bc297d3b621b99a229c3f9a3c0db1ce98573fc5d7098be2603357ba53340e2f"
+  }, # keypair
+  "ae_uat", # network ID
+  "http://localhost:3013/v2", # external URL
```
Comment (thepiwo) [inline README.md]: using testnet here would be good

File: README.md [hunk 13]
```diff
@@ -1,39 +1,45 @@
...
-   balance: 6098700000000000000000, # Can be different
-   id: "ak_1GPPzM3VDKCP5RNEbp2uBNtgGTHRNQmrNkeAKGp7wfPWKYQvM",
-   nonce: 0 # Can be different
- }}
+iex> pubkey = "ak_6A2vcm1Sz6aqJezkLCssUXcyZTX7X8D5UwbuS2fRJr9KkYpRU"
+iex> privkey = "a7a695f999b1872acb13d5b63a830a8ee060ba688a478a08c6e65dfad8a01cd70bb4ed7927f97b51e1bcb5e1340d12335b2a2b12c8bc5221d63c4bcb39d41e61"
+iex> network_id = "ae_uat"
+iex> url = "https://sdk-testnet.aepps.com/v2"
+iex> internal_url = "https://sdk-testnet.aepps.com/v2"
+iex> client = Core.Client.new(%{public: pubkey, secret: privkey}, network_id, url, internal_url)
```
Comment (thepiwo) [inline README.md]: why not reuse the client from the docs above?

File: README.md [hunk 14]
```diff
@@ -1,39 +1,45 @@
...
+iex> source_code = "contract Number =\n  record state = { number : int }\n  function init(x : int) =\n    { number = x }\n  function add_to_number(x : int) = state.number + x"
+iex> init_args = "42"
+iex> Core.Contract.deploy(client, source_code, init_args)
```
Comment (thepiwo) [inline README.md]: doesn't seem to work for me

Comment (thepiwo) [reply README.md]:
```
iex(5)> Core.Contract.deploy(client, source_code, init_args)
** (FunctionClauseError) no function clause matching in Core.Contract.deploy/4    
    
    The following arguments were given to Core.Contract.deploy/4:
```

File: apps/utils/lib/transaction.ex [hunk 15]
```diff
@@ -37,6 +44,182 @@ defmodule Utils.Transaction do
...
+  @default_ttl 0
+
+  @doc """
+  Serialize the list of fields to an RLP transaction binary, sign it with the private key and network ID and post it to the node
+
+  ## Examples
+      iex> connection = AeternityNode.Connection.new("https://sdk-testnet.aepps.com/v2")
+      iex> network_id = "ae_uat"
+      iex> privkey = "a7a695f999b1872acb13d5b63a830a8ee060ba688a478a08c6e65dfad8a01cd70bb4ed7927f97b51e1bcb5e1340d12335b2a2b12c8bc5221d63c4bcb39d41e61"
+      iex> tx = %AeternityNode.Model.ContractCreateTx{
```
Comment (thepiwo) [inline apps/utils/lib/transaction.ex]: seems to be a bad example, where to get this data from?

File: README.md [hunk 16]
```diff
@@ -1,39 +1,45 @@
...
-The return of the function will look like this: 
+**Deploy a smart contract:**
 ``` elixir
-{:ok,
- %AeternityNode.Model.Account{
-   balance: 6098700000000000000000, # Can be different
-   id: "ak_1GPPzM3VDKCP5RNEbp2uBNtgGTHRNQmrNkeAKGp7wfPWKYQvM",
-   nonce: 0 # Can be different
- }}
+iex> pubkey = "ak_6A2vcm1Sz6aqJezkLCssUXcyZTX7X8D5UwbuS2fRJr9KkYpRU"
```
Comment (thepiwo) [inline README.md]: probably a keypair should be generated not hardcoded

File: apps/utils/lib/encoding.ex [hunk 17]
```diff
@@ -22,20 +22,33 @@ defmodule Utils.Encoding do
...
 
   @spec prefix_encode_base58c(String.t(), binary()) :: base58c()
   def prefix_encode_base58c(prefix, payload) when is_binary(payload),
     do: prefix <> "_" <> encode_base58c(payload)
 
-  @doc false
+  @doc """
+  Decode a base58check string to binary and remove its prefix
+
+      iex> Utils.Encoding.prefix_decode_base58c("ak_2XEob1Ub1DWCzeMLm1CWQKrUBsVfF9zLZBDaUXiu6Lr1qLn55n")
```
Comment (thepiwo) [inline apps/utils/lib/encoding.ex]: example tag missing?

File: mix.exs [hunk 18]
```diff
@@ -23,8 +23,16 @@ defmodule AeppSdkElixir.MixProject do
   #
   # Run "mix help deps" for examples and options.
   defp deps do
-    [{:excoveralls, "~> 0.10", only: :test}, {:ex_doc, "~> 0.19", only: :dev, runtime: false}]
+    [
+      {:excoveralls, "~> 0.10", only: :test},
+      {:ex_doc, "~> 0.19", only: :dev, runtime: false},
+      {:aesophia,
```
Comment (thepiwo) [inline mix.exs]: why is this moved to the main mix file not the project specific one?

File: apps/core/lib/contract.ex (hunk 8 above)
Comment (thepiwo) [reply apps/core/lib/contract.ex]: now yields "Invalid Tx"

File: apps/core/lib/contract.ex [hunk 19]
```diff
@@ -0,0 +1,513 @@
...
+  @doc """
+  Deploy a contract
+
+  ## Examples
+      iex> pubkey = "ak_6A2vcm1Sz6aqJezkLCssUXcyZTX7X8D5UwbuS2fRJr9KkYpRU"
+      iex> privkey = "a7a695f999b1872acb13d5b63a830a8ee060ba688a478a08c6e65dfad8a01cd70bb4ed7927f97b51e1bcb5e1340d12335b2a2b12c8bc5221d63c4bcb39d41e61"
+      iex> network_id = "ae_uat"
+      iex> url = "https://sdk-testnet.aepps.com/v2"
+      iex> internal_url = "https://sdk-testnet.aepps.com/v2"
+      iex> client = Core.Client.new(%{pubkey: pubkey, privkey: privkey}, network_id, url, internal_url)
```
Comment (thepiwo) [inline apps/core/lib/contract.ex]: I think the client wont have to be reinitialized in every example, a comment that the client is needed and where this docs can be found should be sufficient

File: mix.exs (hunk 18 above)
Comment (cheezus1) [reply mix.exs]: `aesophia` has `aeserialization` as a dependency and adds it automatically; `aeserialization` is needed in utils

File: README.md (hunk 12 above)
Comment (cheezus1) [reply README.md]: using testnet will make both urls look the same, which can be confusing

File: apps/utils/test/utils_keys_test.exs [hunk 20]
```diff
@@ -63,9 +63,9 @@ defmodule UtilsKeysTest do
 
     %{
       keypair: keypair,
-      pubkey: pubkey,
+      public: pubkey,
       pubkey_binary: pubkey_binary,
-      privkey: privkey,
+      secret: privkey,
```
Comment (thepiwo) [inline apps/utils/test/utils_keys_test.exs]: why is `privkey_binary` not adjusted? I think in general `privatekey` would be the better term

File: README.md [hunk 21]
```diff
@@ -1,39 +1,44 @@
...
-   nonce: 0 # Can be different
- }}
+iex> keypair = Utils.Keys.generate_keypair()
+iex> network_id = "ae_uat"
+iex> url = "https://sdk-testnet.aepps.com/v2"
+iex> internal_url = "https://sdk-testnet.aepps.com/v2"
+iex> client = Core.Client.new(keypair, network_id, url, internal_url)
+iex> source_code = "contract Number =\n  record state = { number : int }\n  function init(x : int) =\n    { number = x }\n  function add_to_number(x : int) = state.number + x"
+iex> init_args = ["42"]
+iex> Core.Contract.deploy(client, source_code, init_args)
```
Comment (DanielaIvanova) [inline README.md]: This is invalid example:

``` elixir
{:error, "Account not found"}
```

File: apps/utils/test/utils_keys_test.exs (hunk 20 above)
Comment (cheezus1) [reply apps/utils/test/utils_keys_test.exs]: can you elaborate ?

Comment (thepiwo) [reply apps/utils/test/utils_keys_test.exs]: privkey was changed to secret, but `privkey_binary` is not changed?

I think in general privatekey would be the better term than `secret`.

Comment (cheezus1) [reply apps/utils/test/utils_keys_test.exs]: we chose `secret` over `privkey` because it differs more with `public`. We can readjust it if you want

File: apps/utils/lib/transaction.ex [hunk 22]
```diff
@@ -37,6 +44,133 @@ defmodule Utils.Transaction do
...
+    await_mining(connection, tx_hash, @await_attempts, type)
+  end
+
+  defp await_mining(_connection, _tx_hash, 0, _type),
+    do:
+      {:error,
+       "Transaction wasn't mined after #{@await_attempts * @await_attempt_interval / 1000} seconds"}
+
+  defp await_mining(connection, tx_hash, attempts, type) do
+    :timer.sleep(@await_attempt_interval)
```
Comment (DanielaIvanova) [inline apps/utils/lib/transaction.ex]: You should use `Process.sleep`, because is better to use `Elixir` functions when it is possible.

File: README.md (hunk 21 above)
Comment (cheezus1) [reply README.md]: @thepiwo suggested that we should use a generated keypair

Comment (thepiwo) [reply README.md]: @DanielaIvanova did you provide an account that exists on testnet?

File: apps/utils/test/utils_keys_test.exs (hunk 20 above)
Comment (thepiwo) [reply apps/utils/test/utils_keys_test.exs]: okay thats a good reason, lets keep it like this
what about the `privkey_binary`?

File: README.md (hunk 21 above)
Comment (DanielaIvanova) [reply README.md]: @thepiwo I just followed their example. Why should I provide something different from an example to see how it works?

Comment (d-velev) [reply README.md]: There's no guarantee that any account will have funds (especially when we're giving the private key as well :smile: )

**Comment by thepiwo** [comment]
@cheezus1 this change does overwrite all fee calculation and doesn't set a higher gas-price as we discussed?
